### PR TITLE
[CBRD-22706] Multi-threaded online index builder

### PR DESCRIPTION
**Pull request by lucianlazar2211**
http://jira.cubrid.org/browse/CBRD-22706

This is the first implementation of the multi-threaded online index builder. The model of work resembles a thread pool which is managed by the index loader. The index loader scans the heap and fetches an object. The index builder then creates the corresponding key and pushes a task to insert it into the index. The thread pool fetches tasks from the queue and begins inserting into the index. Whenever any thread sets an error, the others will stop their execution when they reach the control code and the loading will then be aborted.

### CODE REVIEW

File: src/base/memory_private_allocator.hpp [hunk 1]
```diff
@@ -343,4 +343,26 @@ namespace cubmem
...
+
+template < typename Func, typename...Args >
+void
+switch_to_global_allocator_and_call (Func && func, Args && ... args)
+{
+#if defined(SERVER_MODE)
+  /* Switch to global context. */
+  HL_HEAPID save_id;
+
+  save_id = db_private_set_heapid_to_thread (NULL, 0);
```
Comment (ribram) [inline src/base/memory_private_allocator.hpp]: Use db_change_private_heap instead db_private_set_heapid_to_thread; should help with `#if defined` blocks.

File: src/storage/btree.c [hunk 2]
```diff
@@ -53,6 +53,7 @@
 #include "fault_injection.h"
 #include "dbtype.h"
 #include "thread_manager.hpp"
+#include "thread_entry_task.hpp"
```
Comment (ribram) [inline src/storage/btree.c]: remove unused `thread_entry_task.hpp` header.

File: src/storage/btree_load.c [hunk 3]
```diff
@@ -149,6 +150,38 @@ struct btree_scan_partition_info
   BTID btid;			/* BTID of the current partition. */
 };
 
+struct index_builder_loader_context
```
Comment (ribram) [inline src/storage/btree_load.c]: Move under `// *INDENT-OFF*` block; indent has problems with template types.

File: src/storage/btree_load.c [hunk 4]
```diff
@@ -4757,13 +4790,18 @@ online_index_builder (THREAD_ENTRY * thread_p, BTID_INT * btid_int, HFID * hfids
...
   DB_TYPE single_node_type = DB_TYPE_NULL;
   int attr_offset;
-  DB_VALUE dbvalue;
   DB_VALUE *p_dbvalue;
   int *p_prefix_length;
+  uint64_t tasks_started = 0;
   char midxkey_buf[DBVAL_BUFSIZE + MAX_ALIGNMENT], *aligned_midxkey_buf;
+  index_builder_loader_context load_context;
+  cubthread::entry_workpool * ib_workpool =
+    cubthread::get_manager ()->create_worker_pool (prm_get_integer_value (PRM_ID_INDEX_BUILDER_THREAD_COUNT), 32,
```
Comment (ribram) [inline src/storage/btree_load.c]: Disable indent for statement using namespace prefix. The syntax using `::` confuses it.

File: src/storage/btree_load.c [hunk 5]
```diff
@@ -4852,17 +4897,112 @@ online_index_builder (THREAD_ENTRY * thread_p, BTID_INT * btid_int, HFID * hfids
...
-					   unique_pk, BTREE_OP_ONLINE_INDEX_IB_INSERT, NULL);
-      /* Clear the index key. */
-      pr_clear_value (p_dbvalue);
+      index_builder_loader_task *load_task = new index_builder_loader_task (btid_int->sys_btid, &class_oids[cur_class],
+									    &cur_oid, unique_pk, load_context);
 
-      if (ret != NO_ERROR)
+      load_task->set_key (p_dbvalue);
+
+      cubthread::get_manager ()->push_task (ib_workpool, load_task);
```
Comment (ribram) [inline src/storage/btree_load.c]: thread_get_manager

File: src/storage/btree_load.c [hunk 6]
```diff
@@ -4852,17 +4897,112 @@ online_index_builder (THREAD_ENTRY * thread_p, BTID_INT * btid_int, HFID * hfids
...
+
+      /* Increment tasks started. */
+      tasks_started++;
+
+      /* Check for possible errors. */
+      if (load_context.m_error_code != NO_ERROR)
 	{
-	  return ret;
+	  /* Also stop all threads. */
+	  cubthread::get_manager ()->destroy_worker_pool (ib_workpool);
```
Comment (ribram) [inline src/storage/btree_load.c]: thread_get_manager

File: src/storage/btree_load.c [hunk 7]
```diff
@@ -4852,17 +4897,112 @@ online_index_builder (THREAD_ENTRY * thread_p, BTID_INT * btid_int, HFID * hfids
...
+	}
+    }
+
+  /* Check if the workerpool is empty */
+  do
+    {
+      if (load_context.m_error_code != NO_ERROR)
+	{
+	  /* Also stop all threads. */
+	  cubthread::get_manager ()->destroy_worker_pool (ib_workpool);
```
Comment (ribram) [inline src/storage/btree_load.c]: thread_get_manager

File: src/storage/btree_load.c [hunk 8]
```diff
@@ -4852,17 +4897,112 @@ online_index_builder (THREAD_ENTRY * thread_p, BTID_INT * btid_int, HFID * hfids
...
+	  return load_context.m_error_code;
 	}
 
+      /* Wait for threads to finish. */
+      thread_sleep (10);
     }
+  while (load_context.m_tasks_executed != tasks_started);
+
+  assert (load_context.m_tasks_executed == tasks_started);
+  cubthread::get_manager ()->destroy_worker_pool (ib_workpool);
```
Comment (ribram) [inline src/storage/btree_load.c]: thread_get_manager

File: src/storage/btree_load.c [hunk 9]
```diff
@@ -4852,17 +4897,112 @@ online_index_builder (THREAD_ENTRY * thread_p, BTID_INT * btid_int, HFID * hfids
...
 
   return ret;
 }
+
+index_builder_loader_task::index_builder_loader_task (BTID * btid, OID * class_oid, OID * oid, int unique_pk, index_builder_loader_context & load_context):m_load_context
```
Comment (ribram) [inline src/storage/btree_load.c]: extend indent-off/indent-on block to cover all index_build_loader_task

Comment (ribram) [inline src/storage/btree_load.c]: move ": m_load_context (load_context) to a new line.

File: src/storage/btree_load.c [hunk 10]
```diff
@@ -4852,17 +4897,112 @@ online_index_builder (THREAD_ENTRY * thread_p, BTID_INT * btid_int, HFID * hfids
...
+index_builder_loader_task::~index_builder_loader_task ()
+{
+  switch_to_global_allocator_and_call <int(DB_VALUE *), DB_VALUE * >(pr_clear_value, &m_key);
+}
+// *INDENT-ON*
+
+void
+index_builder_loader_task::set_key (DB_VALUE * key)
+{
+  switch_to_global_allocator_and_call (qdata_copy_db_value, &m_key, key);
```
Comment (ribram) [inline src/storage/btree_load.c]: Please update qdata_copy_db_value: a) return bool, b) src_p signature should be const DB_VALUE *.

Also set_key argument should be `const DB_VALUE * key`

File: src/storage/btree_load.c [hunk 11]
```diff
@@ -4852,17 +4897,112 @@ online_index_builder (THREAD_ENTRY * thread_p, BTID_INT * btid_int, HFID * hfids
...
-      if (ret != NO_ERROR)
+      load_task->set_key (p_dbvalue);
+
+      cubthread::get_manager ()->push_task (ib_workpool, load_task);
+
+      /* Increment tasks started. */
+      tasks_started++;
+
+      /* Check for possible errors. */
+      if (load_context.m_error_code != NO_ERROR)
```
Comment (ribram) [inline src/storage/btree_load.c]: Use m_has_error to check error status of loader. It is more reliable than error code.

File: src/storage/btree_load.c [hunk 12]
```diff
@@ -4852,17 +4897,112 @@ online_index_builder (THREAD_ENTRY * thread_p, BTID_INT * btid_int, HFID * hfids
...
+      if (p_dbvalue != &dbvalue)
+	{
+	  pr_clear_value (p_dbvalue);
+	}
+    }
+
+  /* Check if the workerpool is empty */
+  do
+    {
+      if (load_context.m_error_code != NO_ERROR)
```
Comment (ribram) [inline src/storage/btree_load.c]: m_has_error

File: src/storage/btree_load.c [hunk 13]
```diff
@@ -4852,17 +4897,112 @@ online_index_builder (THREAD_ENTRY * thread_p, BTID_INT * btid_int, HFID * hfids
...
+  switch_to_global_allocator_and_call (qdata_copy_db_value, &m_key, key);
+}
+
+void
+index_builder_loader_task::execute (cubthread::entry & thread_ref)
+{
+  int ret = NO_ERROR;
+
+  /* Check for possible errors set by the other threads. */
+  if (m_load_context.m_error_code != NO_ERROR)
```
Comment (ribram) [inline src/storage/btree_load.c]: m_has_error

File: src/storage/btree_load.c [hunk 14]
```diff
@@ -4852,17 +4897,112 @@ online_index_builder (THREAD_ENTRY * thread_p, BTID_INT * btid_int, HFID * hfids
...
+
+  thread_ref.tran_index = m_load_context.m_tran_index;
+
+  ret = btree_online_index_dispatcher (&thread_ref, &m_btid, &m_key, &m_class_oid, &m_oid,
+				       m_unique_pk, BTREE_OP_ONLINE_INDEX_IB_INSERT, NULL);
+
+  if (ret != NO_ERROR)
+    {
+      /* Set the error. */
+      while (m_load_context.m_has_error.test_and_set (std::memory_order_acquire))
```
Comment (ribram) [inline src/storage/btree_load.c]: Only first error code/message should be registered. That means, only first to change m_has_error flag from false to true (test_and_set returns false) registers the error.

```c
if (m_load_context.m_has_error.test_and_set ())
  {
    // error was already set; register only the first error
  }
else
  {
    // first to set error also must register it
    m_load_context.m_error_code = ret;
    // todo - error message should also be passed.
  }
```

Also, no goto cleanup. Error or no error, m_tasks_executed should be incremented.

File: src/storage/btree_load.c [hunk 15]
```diff
@@ -4852,17 +4897,112 @@ online_index_builder (THREAD_ENTRY * thread_p, BTID_INT * btid_int, HFID * hfids
...
 	{
-	  return ret;
+	  /* Also stop all threads. */
+	  cubthread::get_manager ()->destroy_worker_pool (ib_workpool);
+	  er_set (ER_ERROR_SEVERITY, ARG_FILE_LINE, load_context.m_error_code, 0);
+	  return load_context.m_error_code;
+	}
+
+      /* Clear index key. */
+      if (p_dbvalue != &dbvalue)
```
Comment (ribram) [inline src/storage/btree_load.c]: Should it be cleared if `p_dbvalue == &dbvalue`? Also, probably it should be moved before error case return.

File: src/storage/btree_load.c [hunk 16]
```diff
@@ -149,6 +150,38 @@ struct btree_scan_partition_info
...
+    OID m_class_oid, m_oid;
+    int m_unique_pk;
+    index_builder_loader_context & m_load_context; // Loader context.
+
+public:
+    index_builder_loader_task(BTID * btid, OID * class_oid, OID * oid, int unique_pk,
+                              index_builder_loader_context & load_context);
+    
+    void set_key (DB_VALUE * key);
+
```
Comment (ribram) [inline src/storage/btree_load.c]: remove double empty line.

File: src/storage/btree_load.c [hunk 17]
```diff
@@ -149,6 +150,38 @@ struct btree_scan_partition_info
...
+{
+private:
+    BTID m_btid;
+    DB_VALUE m_key;
+    OID m_class_oid, m_oid;
+    int m_unique_pk;
+    index_builder_loader_context & m_load_context; // Loader context.
+
+public:
+    index_builder_loader_task(BTID * btid, OID * class_oid, OID * oid, int unique_pk,
```
Comment (ribram) [inline src/storage/btree_load.c]: Change constructor arguments to `const BTID *, const OID *, const OID *`.

File: src/storage/btree_load.c [hunk 18]
```diff
@@ -4852,17 +4897,112 @@ online_index_builder (THREAD_ENTRY * thread_p, BTID_INT * btid_int, HFID * hfids
...
+      if (load_context.m_error_code != NO_ERROR)
+	{
+	  /* Also stop all threads. */
+	  cubthread::get_manager ()->destroy_worker_pool (ib_workpool);
+	  er_set (ER_ERROR_SEVERITY, ARG_FILE_LINE, load_context.m_error_code, 0);
+	  return load_context.m_error_code;
 	}
 
+      /* Wait for threads to finish. */
+      thread_sleep (10);
```
Comment (ribram) [inline src/storage/btree_load.c]: After sleep, we should probably also check for interruptions. If a worker hangs, we must allow an user or admin interrupt this job.

File: src/query/execute_schema.c [hunk 19]
```diff
@@ -2959,6 +2961,11 @@ do_create_index (PARSER_CONTEXT * parser, const PT_NODE * statement)
 
   index_name = statement->info.index.index_name ? statement->info.index.index_name->info.name.original : NULL;
 
+  if (statement->info.index.index_status == SM_ONLINE_INDEX_BUILDING_IN_PROGRESS)
+    {
+      ib_thread_count = statement->info.index.ib_threads;
+    }
+
```
Comment (lucianlazar2211) [inline src/query/execute_schema.c]: This was done so that we won't pass the `ib_thread_count` through all the schema functions as a parameter.

**Pull request merged**